src/Ledger/Utxo.lagda

```diff
@@ -83,28 +89,37 @@ minfee : PParams → TxBody → Coin
 minfee pp tx = a * txsize tx + b
   where open PParams pp
 
-certDeposit : DCert → Maybe (Credential × Coin)
-certDeposit (delegate c _ _ v)                 = just $ c , v
-certDeposit (regpool c record { deposit = v }) = just $ c , v
-certDeposit (regdrep c v)                      = just $ c , v
-certDeposit (retirepool _ _)                   = nothing
-certDeposit (deregdrep _)                      = nothing
-certDeposit (ccreghot _ _)                     = nothing
-
-txDeposits : TxBody → Credential ↛ Coin
-txDeposits = foldr _∪⁺_ ∅ᵐ ∘ List.map (f ∘ certDeposit) ∘ txcerts
-  where
-    f : Maybe (Credential × Coin) → Credential ↛ Coin
-    f (just x) = ❴ x ❵ᵐ
-    f nothing = ∅ᵐ
-
--- TODO rewrite this using txDeposits
-totalDeposits : TxBody → Value
-totalDeposits = foldr _+ᵛ_ (inject 0) ∘ List.map (f ∘ certDeposit) ∘ txcerts
-  where
-    f : Maybe (Credential × Coin) → Value
-    f (just (_ , x)) = inject x
-    f nothing = inject 0
+data DepositPurpose : Set where
+  CredentialDeposit  : DepositPurpose
+  PoolDeposit        : DepositPurpose
+  DRepDeposit        : DepositPurpose
+
+instance
+  unquoteDecl DecEq-DepositPurpose = derive-DecEq ((quote DepositPurpose , DecEq-DepositPurpose) ∷ [])
+
+certDeposit : DCert → Maybe (DepositPurpose × Credential × Coin)
+certDeposit (delegate c _ _ v)                 = just (CredentialDeposit , c , v)
+certDeposit (regpool c record { deposit = v }) = just (PoolDeposit       , c , v)
+certDeposit (regdrep c v)                      = just (DRepDeposit       , c , v)
+certDeposit _                                  = nothing
+
+certDepositᵐ : DCert → (DepositPurpose × Credential) ↛ Coin
+certDepositᵐ cert = case certDeposit cert of λ where
+  (just (p , c , v)) → ❴ (p , c) , v ❵ᵐ
+  nothing            → ∅ᵐ
+
+certRefund : DCert → Maybe (DepositPurpose × Credential)
+certRefund (delegate c nothing nothing x) = just (CredentialDeposit , c)
+certRefund (deregdrep c)                  = just (DRepDeposit       , c)
+certRefund _                              = nothing
+
+certRefundˢ : DCert → ℙ (DepositPurpose × Credential)
```

Comment:
I think this can be implemented with `partialToSet`

Reply:
Thank you